File: SpotiFree/MenuController.m
//
//  AppController.m
//  SpotiFree
//
//  Created by Eneas on 21.12.13.
//  Copyright (c) 2013 Eneas. All rights reserved.
//

#import <Sparkle/Sparkle.h>

#import "MenuController.h"
#import "SpotifyController.h"
#import "AppData.h"

@interface MenuController () <NSAlertDelegate, SpotifyControllerDelegate>

@property (unsafe_unretained) IBOutlet NSMenu *statusMenu;
@property (strong) NSStatusItem *statusItem;

@property (unsafe_unretained) IBOutlet NSMenuItem *statusMenuItem;

@property (strong) SpotifyController *spotify;
@property (strong) AppData *appData;

@end

@implementation MenuController

#pragma mark -
#pragma mark Startup
- (void)awakeFromNib {
    BOOL hideMenuBarIcon = [[NSUserDefaults standardUserDefaults] boolForKey:@"hideMenuBarIcon"];
    if (!hideMenuBarIcon) {
        [self setUpMenu];
    }
    
    self.appData = [AppData sharedData];
    
    if ([self.appData isFirstRun]) {
        [self.appData firstRunExecuted];
        
        if (!self.appData.isInLoginItems) {
            NSAlert *alert = [NSAlert alertWithMessageText:@"Do you want Spotifree to run automatically on login?" defaultButton:@"OK" alternateButton:@"No, thanks" otherButton:nil informativeTextWithFormat:@""];
            [alert beginSheetModalForWindow:nil modalDelegate:self didEndSelector:@selector(alertDidEnd:returnCode:contextInfo:) contextInfo:"firstRunAlert"];
        }
    }
    
    [self fixWrongLocationOfScriptingDefinitionFileIfNeeded];
    [[[NSWorkspace sharedWorkspace] notificationCenter] addObserverForName:NSWorkspaceDidLaunchApplicationNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *note) {
        if ([note.userInfo[@"NSApplicationName"] isEqualToString:@"Spotify"]) {
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                [self fixWrongLocationOfScriptingDefinitionFileIfNeeded];
            });
        }
    }];
    
    self.spotify = [SpotifyController spotifyController];
    self.spotify.delegate = self;
    [self.spotify startService];

	[[self.statusMenu itemWithTag:-1] setState:(self.appData.shouldShowNotifications ? NSOnState : NSOffState)];
}

- (BOOL)validateMenuItem:(NSMenuItem *)menuItem {
    if (menuItem.action == @selector(toggleLoginItem:)) {
        menuItem.state = self.appData.isInLoginItems;
    }
    if (menuItem.action == @selector(toggleAutomaticallyChecksForUpdates:)) {
        menuItem.state = [SUUpdater sharedUpdater].automaticallyChecksForUpdates;
    }
    if (menuItem.action == @selector(toggleAutomaticallyDownloadsUpdates:)) {
        menuItem.state = [SUUpdater sharedUpdater].automaticallyDownloadsUpdates;
        return [SUUpdater sharedUpdater].automaticallyChecksForUpdates;
    }
    return YES;
}

- (void)setUpMenu {
    self.statusItem = [[NSStatusBar systemStatusBar] statusItemWithLength:NSSquareStatusItemLength];
    
    NSImage *image = [NSImage imageNamed:@"statusBarIconActiveTemplate"];
    
    [image setTemplate: YES];
    
    [self.statusItem setImage: image];

    [self.statusItem setMenu:self.statusMenu];
    
    [self.statusItem setHighlightMode:YES];
}

- (void)fixWrongLocationOfScriptingDefinitionFileIfNeeded {
    NSFileManager *manager = [NSFileManager defaultManager];
    NSString *spotifyResourceFolder = [[[[NSWorkspace sharedWorkspace] URLForApplicationWithBundleIdentifier:@"com.spotify.client"] path] stringByAppendingString:@"/Contents/Resources/"];
    NSString *rightFile = [spotifyResourceFolder stringByAppendingString:@"Spotify.sdef"];
    
    if ([manager fileExistsAtPath:rightFile])
        return;
    
    NSString *wrongFile = [spotifyResourceFolder stringByAppendingString:@"applescript/Spotify.sdef"];
    
    if ([manager fileExistsAtPath:wrongFile]) {
        [manager copyItemAtPath:wrongFile toPath:rightFile error:nil];
    } else {
        NSAlert *missingAppleScriptDefinitionFileAlert = [NSAlert alertWithMessageText:@"This Spotify version is incompatible with Spotifree" defaultButton:@"Download" alternateButton:@"No, thanks" otherButton:nil informativeTextWithFormat:@"Sorry, but your Spotify app is incompatible with Spotifree. Press \"Download\" to get the latest compatible Spotify version and then restart Spotifree."];
        [missingAppleScriptDefinitionFileAlert beginSheetModalForWindow:nil modalDelegate:self didEndSelector:@selector(alertDidEnd:returnCode:contextInfo:) contextInfo:"missingAppleScriptDefinitionFileAlert"];
        return;
    }
    
    NSRunningApplication *spotify = [[NSRunningApplication runningApplicationsWithBundleIdentifier:@"com.spotify.client"] firstObject];
    if (spotify) {
        NSAlert *alert = [NSAlert alertWithMessageText:@"Spotify restart required" defaultButton:@"OK" alternateButton:@"I'll do it myself" otherButton:nil informativeTextWithFormat:@"Sorry to interrupt, but your Spotify app must be restarted to work with Spotifree. You can do it now or later, manually, if you'd rather enjoy that last McDonald's ad."];
        [alert beginSheetModalForWindow:nil modalDelegate:self didEndSelector:@selector(alertDidEnd:returnCode:contextInfo:) contextInfo:"fixAlert"];
    }
}

#pragma mark -
#pragma SpotifyControllerDelegate
- (void)activeStateShouldGetUpdated:(SFSpotifyState)state {
    if (!self.statusItem)
        return;

	NSString *label;
	NSImage *icon;

	switch (state) {
		case kSFSpotifyStateActive:
			label = @"Active";
			icon = [NSImage imageNamed:@"statusBarIconActiveTemplate"];
			break;
		case kSFSpotifyStateInactive:
			label = @"Inactive";
			icon = [NSImage imageNamed:@"statusBarIconInactiveTemplate"];
			break;
		case kSFSpotifyStateBlockingAd:
			label = @"Muting Ad";
			icon = [NSImage imageNamed:@"statusBarIconBlockingAdTemplate"];
			break;

		default:
			break;
	}

	[self.statusMenuItem setTitle:label];
	[self.statusItem setImage:icon];
}

#pragma mark -
#pragma mark NSAlertModalDelegate
- (void) alertDidEnd:(NSAlert *)alert returnCode:(NSInteger)returnCode contextInfo:(void *)contextInfo {
    if (returnCode == 1) {
        if (strcmp(contextInfo, "firstRunAlert") == 0) {
            [self.appData toggleLoginItem];
        }
        
        if (strcmp(contextInfo, "hideIconAlert") == 0) {
            [self.appData setIsInLoginItems:alert.suppressionButton.state];
            
            [[NSStatusBar systemStatusBar] removeStatusItem:self.statusItem];
            self.statusItem = nil;
            [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"hideMenuBarIcon"];
            [[NSUserDefaults standardUserDefaults] synchronize];
        }
        
        if (strcmp(contextInfo, "fixAlert") == 0) {
            NSRunningApplication *spotify = [[NSRunningApplication runningApplicationsWithBundleIdentifier:@"com.spotify.client"] firstObject];
            [spotify terminate];
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                [[NSWorkspace sharedWorkspace] launchApplication:@"Spotify"];
            });
        }
        
        if (strcmp(contextInfo, "missingAppleScriptDefinitionFileAlert") == 0) {
            [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString: @"http://spotifree.gordinskiy.com/files/Spotify-1.0.1.zip"]];
        }
    }
}

#pragma mark -
#pragma mark IBActions
- (IBAction)aboutItemPushed:(NSMenuItem *)sender {
    [[NSApplication sharedApplication] activateIgnoringOtherApps:YES];
    [[NSApplication sharedApplication] orderFrontStandardAboutPanel:self];
}

- (IBAction)toggleLoginItem:(NSMenuItem *)sender {
    self.appData.isInLoginItems = !self.appData.isInLoginItems;
}

- (IBAction)hideMenuBarIconClicked:(NSMenuItem *)sender {
    NSAlert *alert = [NSAlert alertWithMessageText:@"To show the icon again, simply launch Spotifree from Dock or Finder" defaultButton:@"OK" alternateButton:@"Cancel" otherButton:nil informativeTextWithFormat:@""];
    
    alert.suppressionButton.state = self.appData.isInLoginItems;
    if (!self.appData.isInLoginItems) {
        alert.messageText = [alert.messageText stringByAppendingString:@"\n\nIf you want to make the app truly invisible, we suggest also allowing it to launch at login"];
        alert.suppressionButton.title = @"Run at login";
        [alert setShowsSuppressionButton:YES];
    }
    
    [alert beginSheetModalForWindow:nil modalDelegate:self didEndSelector:@selector(alertDidEnd:returnCode:contextInfo:) contextInfo:"hideIconAlert"];
}

- (IBAction)toggleAutomaticallyChecksForUpdates:(NSMenuItem *)sender {
    [[SUUpdater sharedUpdater] setAutomaticallyChecksForUpdates:![SUUpdater sharedUpdater].automaticallyChecksForUpdates];
    [[SUUpdater sharedUpdater] setAutomaticallyDownloadsUpdates:NO];
}

- (IBAction)toggleAutomaticallyDownloadsUpdates:(NSMenuItem *)sender {
    [[SUUpdater sharedUpdater] setAutomaticallyDownloadsUpdates:![SUUpdater sharedUpdater].automaticallyDownloadsUpdates];
}

- (IBAction)toggleShowNotificationsOnAdBlock:(id)sender {
	BOOL tickStatus = [self.appData toggleShowNotifications];
	[(NSMenuItem*)sender setState:(tickStatus ? NSOnState : NSOffState)];
}

#pragma mark -
#pragma mark Public Methods
- (void)showMenuBarIconIfNeeded {
    if (self.statusItem) {
        return;
    }

    [[NSUserDefaults standardUserDefaults] setBool:NO forKey:@"hideMenuBarIcon"];
    [[NSUserDefaults standardUserDefaults] synchronize];
    [self setUpMenu];
}

@end

File: Spotifree/MenuController.swift
//
//  MenuController.swift
//  Spotifree
//
//  Created by Eneas Rotterdam on 29.12.15.
//  Copyright © 2015 Eneas Rotterdam. All rights reserved.
//

import Cocoa
import Sparkle

class MenuController : NSObject {
    private var statusItem : NSStatusItem?
    
    override init() {
        super.init()
        
        if !DataManager.sharedData.isMenuBarIconHidden() {
            setUpMenu()
        }
    }
    
    func setUpMenu() {
        let statusMenu = NSMenu(title: "Spotifree")
        statusMenu.addItemWithTitle(NSLocalizedString("MENU_INACTIVE", comment: "Spotify state: Inactive"), action: nil, keyEquivalent: "").tag = 1
        statusMenu.addItem(NSMenuItem.separatorItem())
        
        let updateMenu = NSMenu()
        updateMenu.addItemWithTitle(NSLocalizedString("MENU_UPDATES_CHECK_FOR_UPDATES", comment:"Menu: Check For Updates..."), action: #selector(SUUpdater.checkForUpdates(_:)), keyEquivalent: "").target = SUUpdater.sharedUpdater()
        updateMenu.addItem(NSMenuItem.separatorItem())
        updateMenu.addItemWithTitle(NSLocalizedString("MENU_UPDATES_CHECK_AUTOMATICALLY", comment: "Menu: Check Automatically"), action: #selector(MenuController.toggleAutomaticallyCheckForUpdates), keyEquivalent: "").target = self
        updateMenu.addItemWithTitle(NSLocalizedString("MENU_UPDATES_DOWNLOAD_AUTOMATICALLY", comment: "Menu: Download automatically"), action: #selector(MenuController.toggleAutomaticallyDownloadUpdates), keyEquivalent: "").target = self
        let updateItem = NSMenuItem(title:NSLocalizedString("MENU_UPDATES", comment: "Menu: Updates"), action: nil, keyEquivalent: "")
        updateItem.submenu = updateMenu;
        
        statusMenu.addItem(updateItem);
        statusMenu.addItemWithTitle(NSLocalizedString("MENU_HIDE_ICON", comment: "Menu: Hide Icon"), action: #selector(MenuController.hideIconClicked), keyEquivalent: "").target = self
        statusMenu.addItem(NSMenuItem.separatorItem())
        statusMenu.addItemWithTitle(NSLocalizedString("MENU_RUN_AT_LOGIN", comment: "Menu: Run At Login"), action: #selector(MenuController.toggleLoginItem), keyEquivalent: "").target = self
        statusMenu.addItemWithTitle(NSLocalizedString("MENU_NOTIFICATIONS", comment: "Menu: Notifications"), action: #selector(MenuController.toggleNotifications), keyEquivalent: "").target = self
        statusMenu.addItem(NSMenuItem.separatorItem())
        statusMenu.addItemWithTitle(NSLocalizedString("MENU_DONATE", comment: "Menu: Donate"), action: #selector(MenuController.donateLinkClicked), keyEquivalent: "").target = self
        statusMenu.addItemWithTitle(NSLocalizedString("MENU_ABOUT", comment: "Menu: About"), action: #selector(MenuController.aboutItemClicked), keyEquivalent: "").target = self
        statusMenu.addItemWithTitle(NSLocalizedString("MENU_QUIT", comment: "Menu: Quit"), action: "terminate:", keyEquivalent: "q");
        statusMenu.addItem(NSMenuItem.separatorItem())
        
        statusItem = NSStatusBar.systemStatusBar().statusItemWithLength(NSSquareStatusItemLength)
        statusItem!.image = NSImage(named: "statusBarIconInactiveTemplate")
        statusItem!.menu = statusMenu
        statusItem!.highlightMode = true
    }
    
    override func validateMenuItem(menuItem: NSMenuItem) -> Bool {
        if menuItem.action == #selector(MenuController.toggleNotifications) {
            menuItem.state = Int(DataManager.sharedData.shouldShowNofifications())
        }
        if menuItem.action == #selector(MenuController.toggleLoginItem) {
            menuItem.state = Int(DataManager.sharedData.isInLoginItems())
        }
        if menuItem.action == #selector(MenuController.toggleAutomaticallyCheckForUpdates) {
            menuItem.state = Int(SUUpdater.sharedUpdater().automaticallyChecksForUpdates)
        }
        if menuItem.action == #selector(MenuController.toggleAutomaticallyDownloadUpdates) {
            menuItem.state = Int(SUUpdater.sharedUpdater().automaticallyDownloadsUpdates)
            return SUUpdater.sharedUpdater().automaticallyChecksForUpdates
        }
        return true
    }
    
    func hideIconClicked() {
        let alert = NSAlert()
        alert.messageText = NSLocalizedString("ALERT_HIDE_ICON_INFO", comment: "Alert info: To show the icon again, simply launch Spotifree from Dock or Finder")
        alert.addButtonWithTitle(NSLocalizedString("OK", comment: "General: OK"))
        alert.addButtonWithTitle(NSLocalizedString("CANCEL", comment: "General: Cancel"))
        
        if !DataManager.sharedData.isInLoginItems() {
            alert.informativeText = NSLocalizedString("ALERT_HIDE_ICON_LAUNCH_AT_LOGIN_INFO", comment: "Alert info: If you want to make the app truly invisible, we suggest also allowing it to launch at login")
            alert.showsSuppressionButton = true
            alert.suppressionButton?.title = NSLocalizedString("MENU_RUN_AT_LOGIN", comment: "Menu: Run At Login")
            alert.suppressionButton?.state = NSOffState
        }
        
        statusItem?.highlightMode = false
        let response = alert.runModal()
        if response == NSAlertFirstButtonReturn {
            DataManager.sharedData.setMenuBarIconHidden(true)
            NSStatusBar.systemStatusBar().removeStatusItem(statusItem!)
            statusItem = nil
            
            if alert.suppressionButton?.state == NSOnState {
                DataManager.sharedData.toggleLoginItem()
            }
        }
        statusItem?.highlightMode = true
    }
    
    func showMenuBarIconIfNeeded() {
        if self.statusItem != nil {return}
        
        DataManager.sharedData.setMenuBarIconHidden(false)
        setUpMenu()
    }
    
    func toggleNotifications() {
        DataManager.sharedData.toggleShowNotifications()
    }
    
    func toggleLoginItem() {
        DataManager.sharedData.toggleLoginItem()
    }
    
    func toggleAutomaticallyCheckForUpdates() {
        SUUpdater.sharedUpdater().automaticallyChecksForUpdates = !SUUpdater.sharedUpdater().automaticallyChecksForUpdates
        SUUpdater.sharedUpdater().automaticallyDownloadsUpdates = false;
    }
    
    func toggleAutomaticallyDownloadUpdates() {
        SUUpdater.sharedUpdater().automaticallyDownloadsUpdates = !SUUpdater.sharedUpdater().automaticallyDownloadsUpdates
    }
    
    func donateLinkClicked() {
        let donateURL = "https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=UG7ECWW2QNWBJ&lc=US&item_name=Donation%20for%20the%20development%20of%20Spotifree%20app&currency_code=USD&bn=PP%2dDonationsBF%3abtn_donateCC_LG%2egif%3aNonHosted"
        
        NSWorkspace.sharedWorkspace().openURL(NSURL(string: donateURL)!)
    }
    
    func aboutItemClicked() {
        NSApplication.sharedApplication().activateIgnoringOtherApps(true)
        NSApplication.sharedApplication().orderFrontStandardAboutPanel(self)
    }
}

extension MenuController : SpotifyManagerDelegate {
    func spotifreeStateChanged(state: SFSpotifreeState) {
        if let statusItem = statusItem {
            var label = "Status Unknown"
            var icon : NSImage?
            
            switch state {
            case .Active:
                label = NSLocalizedString("MENU_ACTIVE", comment: "Spotify state: Active")
                icon = NSImage(named: "statusBarIconActiveTemplate")
            case .Muting:
                label = NSLocalizedString("MENU_MUTING_AD", comment: "Spotify state: Muting Ad")
                icon = NSImage(named: "statusBarIconBlockingAdTemplate")
            case .Inactive:
                label = NSLocalizedString("MENU_INACTIVE", comment: "Spotify state: Inactive")
                icon = NSImage(named: "statusBarIconInactiveTemplate")
            }
            
            statusItem.image = icon
            statusItem.menu?.itemWithTag(1)?.title = label
        }
    }
}
